File: partiql-plan/src/main/java/org/partiql/plan/rex/RexSpread.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;
import org.partiql.spi.types.PType;

import java.util.List;

/**
 * Logical spread expression abstract base class.
 */
public abstract class RexSpread extends RexBase {

    /**
     * @return new RexSpread instance
     */
    @NotNull
    public static RexSpread create(@NotNull List<Rex> args) {
        return new Impl(args);
    }

    /**
     * @return list of spread arguments (the operands)
     */
    public abstract List<Rex> getArgs();

    @NotNull
    @Override
    protected final RexType type() {
        return RexType.of(PType.struct());
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.vararg(getArgs());
        return List.of(c0);
    }

    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitSpread(this, ctx);
    }

    private static class Impl extends RexSpread {

        private final List<Rex> args;

        private Impl(List<Rex> args) {
            this.args = args;
        }

        @Override
        public List<Rex> getArgs() {
            return args;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexSpread.kt
package org.partiql.plan.rex

import org.partiql.plan.Visitor

/**
 * TODO DOCUMENTATION
 */
public interface RexSpread : Rex {

    public fun getArgs(): List<Rex>

    override fun getChildren(): Collection<Rex> = getArgs()

    override fun <R, C> accept(visitor: Visitor<R, C>, ctx: C): R = visitor.visitSpread(this, ctx)
}

/**
 * Default [RexSpread] operator intended for extension.
 */
internal class RexSpreadImpl(args: List<Rex>, type: RexType) : RexSpread {

    // DO NOT USE FINAL
    private var _args = args
    private var _type = type

    override fun getArgs(): List<Rex> = _args

    override fun getType(): RexType = _type

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RexSpreadImpl) return false
        if (_args != other._args) return false
        if (_type != other._type) return false
        return true
    }

    override fun hashCode(): Int {
        var result = _args.hashCode()
        result = 31 * result + _type.hashCode()
        return result
    }
}
